client/data/home/use-home-layout-query.ts

```diff
@@ -0,0 +1,46 @@
+/**
+ * External dependencies
+ */
+import { useQuery, UseQueryResult } from 'react-query';
+
+/**
+ * Internal dependencies
+ */
+import wpcom from 'calypso/lib/wp';
+import { useHomeLayoutQueryParams } from './use-home-layout-query-params';
+
+interface Options {
+	enabled?: boolean;
+}
+
+const useHomeLayoutQuery = (
+	siteId: number | null,
+	{ enabled = true }: Options = {}
+): UseQueryResult => {
+	const query = useHomeLayoutQueryParams();
+
+	return useQuery(
+		// Don't include `dev` or `view` in query key, we want use of the hook to
+		// have the same idea of what the current view is, regardless of dev flags.
+		[ 'home-layout', siteId ],
+		() =>
+			wpcom.req.get(
+				{
+					path: `/sites/${ siteId }/home/layout`,
+					apiNamespace: 'wpcom/v2',
+				},
+				query
+			),
+		{
+			enabled: !! siteId && enabled,
+
+			// The `/layout` endpoint can return a random view. Disable implicit refetches
+			// so the view doesn't change without some user action.
+			staleTime: Infinity,
```

Comment:
Are there any conditions for which we'd want to re-enable re-fetching? For example, if another user has logged in (and uses the same site), or if the locale has changed, or something else, non-site related? Perhaps it's just fine if we keep the same view for the next user who logged in?

Reply:
The query cache gets destroyed on page reload, and logging in as another user always involves a full reload. Calypso can't switch a user at runtime. Even if the cache starts being persisted to local storage, it will be keyed by user ID.

Reply:
> Are there any conditions for which we'd want to re-enable re-fetching? For example, if another user has logged in (and uses the same site)

Really interesting question about how resilient we think hooks/components should be. It made me think about what we'd do if we wanted our components to render in storybook, would we have to assume that the Calypso layout wasn't there to do a bunch of the work for us? 🤔 I imagine in storybook we'd clear the cache between navigating to different stories.
I think this is similar to how in most Calypso code it's ok to assume the selected site data is available in Redux, even though to be fully resilient you could argue each component should request it just to be sure.

Just assuming that the query cache will be destroyed during logout feels like the right trade off to me.

Reply:
I see the `/home/layout` endpoint being a special case because we want to have a very precise control over when we fetch from the endpoint. Because the endpoint returns randomized results, we want to fetch from it once and only once for each Home pageview. Re-fetching, which is usually a harmless way to get fresh data, is undesired here.

The default React Query behavior is to fetch the data on i) mount ii) window focus iii) reconnect. Then mark it immediately as stale (`staleTime = 0`) and refetch on the next mount/focus/reconnect. But for `/home/layout`, we want to do the fetch only on one specific mount and nowhere else.

A possible improvement of the current situation, to make sure that we fetch only once on one specific mount, would be to call `useHomeLayoutQuery` only at that one place, and drill down the current `layout` down as a prop or in a context.